ssl-config-akka/src/main/scala/com/typesafe/sslconfig/akka/io/NegotiateNewSession.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package com.typesafe.sslconfig.akka.io
+
+import javax.net.ssl.SSLParameters
+
+import com.typesafe.sslconfig.ssl.ClientAuth
+
+import scala.annotation.varargs
+import scala.collection.immutable
+
+/**
+ * Initiate a new session negotiation. Any `SendBytes` commands following
+ * this one will be held back (i.e. back-pressured) until the new handshake is
+ * completed, meaning that the bytes following this message will be encrypted
+ * according to the requirements outlined here.
+ *
+ * Each of the values in this message is optional and will have the following
+ * effect if provided:
+ *
+ *  - `enabledCipherSuites` will be passed to `SSLEngine::setEnabledCipherSuites()`
+ *  - `enabledProtocols` will be passed to `SSLEngine::setEnabledProtocols()`
+ *  - `clientAuth` will be passed to `SSLEngine::setWantClientAuth()` or `SSLEngine.setNeedClientAuth()`, respectively
+ *  - `sslParameters` will be passed to `SSLEngine::setSSLParameters()`
+ */
+case class NegotiateNewSession(
+    enabledCipherSuites: Option[immutable.Seq[String]],
+    enabledProtocols: Option[immutable.Seq[String]],
+    clientAuth: ClientAuth,
+    sslParameters: Option[SSLParameters]) { // extends SslTlsOutbound {
+
+  /**
+   * Java API: Make a copy of this message with the given `enabledCipherSuites`.
+   */
+  @varargs
+  def withCipherSuites(s: String*) = copy(enabledCipherSuites = Some(s.toList))
+
+  /**
+   * Java API: Make a copy of this message with the given `enabledProtocols`.
+   */
+  @varargs
+  def withProtocols(p: String*) = copy(enabledProtocols = Some(p.toList))
+
+  /**
+   * Java API: Make a copy of this message with the given [[ClientAuth]] setting.
+   */
+  def withClientAuth(ca: ClientAuth) = copy(clientAuth = ca)
+
+  /**
+   * Java API: Make a copy of this message with the given [[SSLParameters]].
+   */
+  def withParameters(p: SSLParameters) = copy(sslParameters = Some(p))
+}
+
+object NegotiateNewSession extends NegotiateNewSession(None, None, ClientAuth.Default, None) {
+  /**
+   * Java API: obtain the default value (which will leave the SSLEngine’s
+   * settings unchanged).
+   */
+  def withDefaults = this
+}
```

Comment:
This is moved here from akka.stream.io, otherwise we'd have to dance around and with making ssl-config-akka depend on akka-stream :( This makes it impossible to extend `SslTlsOutbound`